core/vtk/ttkConnectedComponents/ttkConnectedComponents.cpp

```diff
@@ -0,0 +1,170 @@
+#include <ttkConnectedComponents.h>
+
+#include <vtkDataArray.h>
+#include <vtkInformation.h>
+#include <vtkObjectFactory.h>
+#include <vtkPointData.h>
+#include <vtkPolyData.h>
+#include <vtkSmartPointer.h>
+
+#include <vtkFloatArray.h>
+#include <vtkIntArray.h>
+
+#include <ttkMacros.h>
+#include <ttkUtils.h>
+
+vtkStandardNewMacro(ttkConnectedComponents);
+
+ttkConnectedComponents::ttkConnectedComponents() {
+  this->SetNumberOfInputPorts(1);
+  this->SetNumberOfOutputPorts(2);
+
+  // Suppress warning if one does not set the optional input array
+  this->SetInputArrayToProcess(0, 0, 0, 0, "%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%");
+}
+
+ttkConnectedComponents::~ttkConnectedComponents() {
+}
+
+int ttkConnectedComponents::FillInputPortInformation(int port,
+                                                     vtkInformation *info) {
+  if(port == 0) {
+    info->Set(vtkAlgorithm::INPUT_REQUIRED_DATA_TYPE(), "vtkDataSet");
```

Comment:
not sur about this, only dataset supported by the ttkTriangulation are supported no ?

Reply:
`vtkDataSet` includes `vtkImageData` and `vtkUnstructuredGrid` which are both supported by the triangulation.

Reply:
The problem is that it also includes types that are not (`vtkStructuredGrid` for example). When using the VTK module, this may lead to a TTK filter being called on a bad dataset and a crash. We may call the `ttkTriangulation::FillInputPortInformation` for the case where `port == 0` to keep the type checking clean.

Reply:
Yes, but this is a problem all over the place. Theoretically we would have to check if the input is either a `vtkPolyData`, `vtkUnstructuredGrid`, or `vtkImageData` object. But in all filter so far we just check for `vtkDataSet` and return an error if it is not possible to derive a triangulation. If we decide to fix this I would suggest doing it in an own PR that fixes this across the board, But I would just keep it like this.

Reply:
all right :)

Reply:
`core/vtk/ttkAlgorithm/ttkTriangulationFactory.cpp` has various checks implemented and report warning messages on the standard output if the input is not supported (for instance: unsupported class [NOT: vtkImageData, vtkPolytData, vtkUnstructuredGrid], cells are not simplices, cells have non-uniform dimensions, etc.).
I believe this is a sufficient safeguard for the moment, isn't it?

Reply:
I think this is a sufficient safeguard, as it should not lead to a crash. If my understanding is correct, these checks are called on the `RequesetData` pass. Filling the `InputPortInformation` with the valid datasets only may allow the pipeline for an early exit with a clear message.

Reply:
you can try to the following to see the safeguard in action:
```
from paraview.simple import *
wavelet1 = Wavelet(registrationName='Wavelet1')
cleantoGrid1 = CleantoGrid(registrationName='CleantoGrid1', Input=wavelet1)
tTKScalarFieldCriticalPoints1 = TTKScalarFieldCriticalPoints(registrationName='TTKScalarFieldCriticalPoints1', Input=cleantoGrid1)
UpdatePipeline(time=0.0, proxy=tTKScalarFieldCriticalPoints1)
```

Reply:
I also think it is safe, but I can see Charle's point that it would be nice that the ttk filters would only accept triangulable input data objects. Like I said this can be achieved by replacing `info->Set(vtkAlgorithm::INPUT_REQUIRED_DATA_TYPE(), "vtkDataSet");` with the checks for the supported three data types:

`
info->Remove(vtkAlgorithm::INPUT_REQUIRED_DATA_TYPE());
info->Append(vtkAlgorithm::INPUT_REQUIRED_DATA_TYPE(), "vtkImageData");
info->Append(vtkAlgorithm::INPUT_REQUIRED_DATA_TYPE(), "vtkPolyData");
info->Append(vtkAlgorithm::INPUT_REQUIRED_DATA_TYPE(), "vtkUnstructuredGrid");
`

Reply:
you should get the following output:
```
[ScalarFieldCriticalPoints] %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
[TriangulationFactory] [WARNING] Cells are not simplices.
[TriangulationFactory] [WARNING] Consider using `vtkTetrahedralize` in pre-processing.
[ScalarFieldCriticalPoints] [ERROR] Unable to retrieve/initialize triangulation for 'vtkUnstructuredGrid'
(  17.889s) [paraview        ]       vtkExecutive.cxx:752    ERR| vtkPVCompositeDataPipeline (0x5571da9ea770): Algorithm ttkScalarFieldCriticalPoints(0x5571d877d4b0) returned failure for request: vtkInformation (0x5571d8d6e530)
  Debug: Off
  Modified Time: 1319979
  Reference Count: 1
  Registered Events: (none)
  Request: REQUEST_DATA
  FROM_OUTPUT_PORT: 0
  ALGORITHM_AFTER_FORWARD: 1
  FORWARD_DIRECTION: 0
```

Reply:
> I also think it is safe, but I can see Charle's point that it would be nice that the ttk filters would only accept triangulable input data objects. Like I said this can be achieved by replacing `info->Set(vtkAlgorithm::INPUT_REQUIRED_DATA_TYPE(), "vtkDataSet");` with the checks for the supported three data types:
> 
> ```
> info->Remove(vtkAlgorithm::INPUT_REQUIRED_DATA_TYPE());
> info->Append(vtkAlgorithm::INPUT_REQUIRED_DATA_TYPE(), "vtkImageData");
> info->Append(vtkAlgorithm::INPUT_REQUIRED_DATA_TYPE(), "vtkPolyData");
> info->Append(vtkAlgorithm::INPUT_REQUIRED_DATA_TYPE(), "vtkUnstructuredGrid");
> ```

that only partially solves the issue as vtkPolyData and vtkUnstructuredGrid can still represent some invalid inputs for TTK, for instance cells which are not simplices (e.g. quads) or with non-uniform dimensions (e.g. mixture of edges and triangles).

Reply:
@julien-tierny this check is nice, and for the manifold / simplices check we have no choice but to do them in Request Data as we need the data set to be available :)

@jonas this is indeed what we want, but this would make sense at a TTK wide level. So I would argue to keep this filter TTK compliant and let this change as a futur work